### DISCUSSION TITLE
Example Lark grammars

### DISCUSSION THREAD

**Discussion by kaby76** [lark-parser/lark#924] [General]
Are there any examples of non-trivial programming language grammars for Lark, e.g., Java, C#, Python, Go, Dart, Scala, C++, Sql, etc?

**Comment by MegaIng** [comment]
Yes. In the [`examples/advanced` directory](https://github.com/lark-parser/lark/tree/master/examples/advanced) there is a python grammar (And a few other examples, but those grammars are not that complex).

**Reply by kaby76** [reply]
Excellent. Are there any others besides the Python grammars, again, Java, C#, etc.?

**Reply by MegaIng** [reply]
Not in any official way. You might be able to find some by searching projects using lark, but in general it should be generally pretty straightforward to translate a different EBNF syntax to lark syntax..

**Reply by kaby76** [reply]
I am writing a convertor to/from Antlr/Lark. I'm not familiar with Lark, so I'm looking for realistic examples to analyze carefully along with the code. The Python3 grammar is a good start because it is very similar to the one in antlr/grammars-v4.

**Comment by erezsh** [comment]
Sorry, nothing comes to mind, other than the Python grammar in the examples.

Relevant thread? https://github.com/lark-parser/lark/issues/333

**Reply by kaby76** [reply]
Syntactically, at the parser-grammar level, Lark and Antlr look similar. Antlr too can parse right-recursive grammars, but it is best LL. But, this is not an issue because I have a refactoring system that can convert recursion/ebnf to any form (right, left, ebnf, mutual recursion).

However, I do see that there is a gap in the lexer rules, e.g.,

    from lark import Lark
    l = Lark('''
    start : (A | B)+
    A : /a|ab/          // or A: 'a' | 'ab'
    B : /b/
             ''')
    print( l.parse("aba") )

versus

	grammar A;
	start: (A | B)+ EOF;
	A : 'a' | 'ab' ;
	B : 'b' ;

These recognize different tokens for "aba" (Lark: 'a', 'b', 'a'; Antlr: 'ab', a). Antlr follows (1) longest lexer rule match, (2) if there are multiple lexer rules that match, always choose the first. Lark r.e.'s selects the first matching rule. I don't know how much this will be a problem for the [grammars-v4](https://github.com/antlr/grammars-v4) grammars.

And, yes, semantic predicates appear in a significant number of the grammars. In the case of Java, the predicates (for lexing identifiers) can be replaced with non-predicate code. I expect the non-"optimized" grammar for Java to work well here.

**Reply by erezsh** [reply]
> I have a refactoring system that can convert recursion/ebnf to any form (right, left, ebnf, mutual recursion).

That's very interesting. I would like to hear more!

> These recognize different tokens for "aba"

When ambiguity is involved, it's worth noting that Earley and LALR might produce different results too. (and your example uses the default, which is Earley)

However,

- In Earley, it's possible to add priority to rules to guide disambiguation, or get the SPPF and disambiguate manually.
- In LALR, it's possible to add priority to terminals, and that way force the order of regex evaluation (and also prevent "keyword resolution", aka the "unless" mechanism)

**Reply by kaby76** [reply]
My grammar toolkit is called, apropos, "Trash". The source is [in Github](https://github.com/kaby76/Domemtech.Trash). It's tied tightly to Antlr, but with another rewrite, I plan to have a different IR.

**Reply by erezsh** [reply]
Looks like a worthy project. Best of luck!

And let us know if there's anything we can do to help.

**Reply by kaby76** [reply]
A couple of translations:

[grammars-v4/abb/*.g4](https://github.com/antlr/grammars-v4/tree/master/abb) to [abb.lark](https://gist.github.com/kaby76/f50080dd5fdcb5ffdabbe70b44e4debe)

[lark.lark](https://github.com/lark-parser/lark/blob/d676df9b888ead42daffd31c035d95241bff0920/lark/grammars/lark.lark) to [LarkLexer.g4 and LarkParser.g4](https://gist.github.com/kaby76/8a96f5b79348b5006d020fd9f8255d72)

The translation code is very rudimentary and the output requires hand tweaking for the moment. I am planning to extend StringTemplates to work with trees and create a stand-alone template tool. For the moment, I'm mostly just codifying translation rules.

**Reply by erezsh** [reply]
Looks like a nice start! A few comments:

- `BOOL : B O O L ` can just be written as `BOOL: "bool"i`. And of course, many of the terminals don't need to be declared. So an idiomatic grammar would be much smaller, half the size or even less.
- It's a better style if the `+` operator doesn't have a preceding space.

**Reply by kaby76** [reply]
Yes, I have an optimization already for Antlr4 to add/remove case-insensitive string literals, e.g., "A : 'a';" <=> "A : [aA];". Intertoken spacing is preserved in the usual case, but I have transformations for reformating. I'm not sure how template translation will work with inter-token spacing, especially comments. But, it is important: Antlr allows ':' in a rule to be on a separate line from the LHS symbol; Lark requires it to be on the same line. I am sure there will be other optimizations needed.
